### PR TITLE
Fixup categoryorder for missing values in cartesian traces

### LINKED ISSUE #5267: category order bug
This pen https://codepen.io/nicolaskruchten/pen/OJXrOLG shows (on my machine anyway) the categories being out of order:

image

reported in https://github.com/plotly/plotly.py/issues/2350 but oddly not always reproducible.

### PR DESCRIPTION
**Pull request by archmoj**
Fixes #5267.

@plotly/plotly_js 
cc: @cldougl

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
@antoinerg can you take a look here as well plz, given that you did the original implementation here?

The original reports were strangely hard to reproduce: https://github.com/plotly/plotly.js/issues/5267 ... does your fix sort of suggest why this might be machine/browser/environment-dependent, @archmoj ?

**Comment by archmoj** [comment]
> The original reports were strangely hard to reproduce: #5267 ... does your fix sort of suggest why this might be machine/browser/environment-dependent, @archmoj ?

Possibly related to the fact that the `width` and `height` were not set on the mock you were testing...

**Comment by nicolaskruchten** [comment]
Ah that's an interesting possibility... But why does the category order depend on the width?!?

**Comment by archmoj** [comment]
> Ah that's an interesting possibility... But why does the category order depend on the width?!?

I've not noticed this bug on different browsers.
Anyway the width may impact the bin size for histograms maybe?

**Comment by nicolaskruchten** [comment]
This is a categorical histogram so there is no bin size :)

**Comment by antoinerg** [comment]
@archmoj can you point me where in the code the existing behavior was changed.

**Comment by archmoj** [comment]
@nicolaskruchten 
Testing [this mock](https://codepen.io/MojtabaSamimi/pen/yLJGPYJ?editors=0010) on different OS+browsers leads to the identical error as far as I could test :)

File: src/plots/plots.js [hunk 1]
```diff
@@ -3181,16 +3180,19 @@ function sortAxisCategoriesByValue(axList, gd) {
...
+                        if(value === undefined) value = cdi.v;
+                        if(value === undefined) value = isX ? cdi.y : cdi.x;
+
+                        if(!Array.isArray(value)) {
+                            if(value === undefined) value = [];
+                            else value = [value];
                         }
-                        if(!Array.isArray(value)) value = [value];
                         for(l = 0; l < value.length; l++) {
-                            categoriesValue[cat][1].push(value[l]);
```
Comment (antoinerg) [inline src/plots/plots.js:3193]: `cat` was ill-defined here, wasn't it?

File: src/plots/plots.js [hunk 2]
```diff
@@ -3181,16 +3180,19 @@ function sortAxisCategoriesByValue(axList, gd) {
                         }
                     } else {
                         // For all other 2d cartesian traces
-                        if(axLetter === 'x') {
-                            cat = cdi.p + 1 ? cdi.p : cdi.x;
-                            value = cdi.s || cdi.v || cdi.y;
```
Comment (archmoj) [inline src/plots/plots.js:3186]: @antoinerg please note `cdi.s` here could be 0 and it is a valid value although it is zero.
We should have not used `cdi.v` or `cdi.(x|y)` instead in that case.

Comment (antoinerg) [reply src/plots/plots.js:3186]: Nice catch!

**Comment by antoinerg** [comment]
Thanks @archmoj for the nice code cleanup and bug fix! LGTM :woman_dancing:

File: src/plots/plots.js (hunk 1 above)
Comment (archmoj) [reply src/plots/plots.js:3193]: No. It was defined. This is just the result of refactoring to drop the extra variable `cat`.

**Pull request merged**